phpBB/install/startup.php

```diff
@@ -103,17 +97,129 @@ function installer_msg_handler($errno, $msg_text, $errfile, $errline)
 	return false;
 }
 
+/**
+ * Register class loaders for installer
+ *
+ * @param string $phpbb_root_path phpBB root path
+ * @param string $phpEx PHP file extension
+ */
+function installer_class_loader($phpbb_root_path, $phpEx)
+{
+	$phpbb_class_loader_new = new \phpbb\class_loader('phpbb\\', "{$phpbb_root_path}install/update/new/phpbb/", $phpEx);
+	$phpbb_class_loader_new->register();
+	$phpbb_class_loader = new \phpbb\class_loader('phpbb\\', "{$phpbb_root_path}phpbb/", $phpEx);
+	$phpbb_class_loader->register();
+	$phpbb_class_loader = new \phpbb\class_loader('phpbb\\convert\\', "{$phpbb_root_path}install/convert/", $phpEx);
+	$phpbb_class_loader->register();
+	$phpbb_class_loader_ext = new \phpbb\class_loader('\\', "{$phpbb_root_path}ext/", $phpEx);
+	$phpbb_class_loader_ext->register();
+}
+
+/**
+ * Installer shutdown function. Tries to resolve errors that might have occured
+ * during execution of installer
+ *
+ * @param int $display_errors Original display errors value
+ */
+function installer_shutdown_function($display_errors)
+{
+	$error = error_get_last();
+
+	if ($error)
+	{
+		// Restore original display errors value
+		@ini_set('display_errors', $display_errors);
+
+		// Manually define phpBB root path and phpEx. These will not be passed
+		// on from app.php
+		$phpbb_root_path = __DIR__ . '/../';
+		$phpEx = 'php';
+
+		installer_class_loader($phpbb_root_path, $phpEx);
+		$supported_error_levels = E_ALL & ~E_NOTICE & ~E_DEPRECATED & ~E_USER_DEPRECATED;
+
+		$cache = new \phpbb\cache\driver\file(__DIR__ . '/../cache/installer');
+		$filesystem = new \phpbb\filesystem\filesystem();
+		if (strpos($error['file'], $filesystem->realpath($cache->cache_dir)) !== false)
+		{
+			$file_age = @filemtime($error['file']);
+
+			if ($file_age !== false && ($file_age + 60) < time())
+			{
+				$cache->purge();
+
+				$symfony_request = new \phpbb\symfony_request(new \phpbb\request\request(new \phpbb\request\type_cast_helper()));
+
+				header('Location: ' . $symfony_request->getRequestUri());
+				exit();
+			}
+			else
+			{
+				// Language system is not available
+				die('The installer has detected an issue with a cached file. Try reloading the page and/or manually clearing the cache to resolve the issue. If you require further assistance, please visit the <a href="https://www.phpbb.com/community/">phpBB support forums</a>.');
```

Comment:
I am usually unbiased towards opening links in a new window or not, but I think this one deserves a consideration for a `target="_blank"`.